File: frontend/src/utils/pricing.js
/**
 * Client-side pricing calculation (mirrors backend algorithms/pricing.py).
 */

const MIN_PRICE = 500
const MAX_PRICE_RATIO = 0.90
const DEFAULT_FUEL_COST_PER_KM = 35
const DEFAULT_PLATFORM_FEE = 100

export function calculatePrice(
  baseTripCost,
  truckCapacityKg,
  shipmentWeightKg,
  detourDistanceKm,
  fuelCostPerKm = DEFAULT_FUEL_COST_PER_KM,
  platformFee   = DEFAULT_PLATFORM_FEE,
) {
  if (!truckCapacityKg || truckCapacityKg <= 0) truckCapacityKg = 10000

  const weightRatio  = Math.min(1, shipmentWeightKg / truckCapacityKg)
  const volumeShare  = baseTripCost * weightRatio
  const detourCost   = detourDistanceKm * fuelCostPerKm
  let   finalPrice   = volumeShare + detourCost + platformFee

  finalPrice = Math.max(MIN_PRICE, finalPrice)
  finalPrice = Math.min(finalPrice, baseTripCost * MAX_PRICE_RATIO)

  const dedicatedTruckCost = baseTripCost
  const estimatedSavings   = dedicatedTruckCost - finalPrice
  const savingsPercent     = dedicatedTruckCost > 0
    ? +((estimatedSavings / dedicatedTruckCost) * 100).toFixed(1)
    : 0

  const weightPct = (weightRatio * 100).toFixed(1)
  const explanation = (
    `Your cargo occupies ${weightPct}% of the truck's capacity. ` +
    `You pay a proportional share (₹${Math.round(volumeShare).toLocaleString('en-IN')}) ` +
    `plus ₹${Math.round(detourCost).toLocaleString('en-IN')} for the ${detourDistanceKm.toFixed(1)} km detour ` +
    `and a ₹${platformFee.toLocaleString('en-IN')} platform fee. ` +
    `This saves you ₹${Math.round(estimatedSavings).toLocaleString('en-IN')} vs a dedicated truck.`
  )

  return {
    volumeShare:       +volumeShare.toFixed(2),
    detourCost:        +detourCost.toFixed(2),
    platformFee:       +platformFee.toFixed(2),
    finalPrice:        +finalPrice.toFixed(2),
    dedicatedTruckCost:+dedicatedTruckCost.toFixed(2),
    estimatedSavings:  +estimatedSavings.toFixed(2),
    savingsPercent,
    weightRatio:       +weightRatio.toFixed(4),
    explanation,
  }
}

export function formatINR(amount) {
  return `₹${Math.round(amount).toLocaleString('en-IN')}`
}

File: backend/algorithms/pricing.py
"""
Pricing Algorithm
=================

Formula:
  volumeShare  = baseTripCost × (shipmentWeightKg / truckCapacityKg)
  detourCost   = detourDistanceKm × fuelCostPerKm
  finalPrice   = volumeShare + detourCost + platformFee

  dedicatedTruckCost = baseTripCost    (what shipper would pay alone)
  estimatedSavings   = dedicatedTruckCost − finalPrice

Safeguards:
  - finalPrice is always >= MIN_PRICE_INR (₹500)
  - finalPrice is always <= dedicatedTruckCost × MAX_PRICE_RATIO (90%)
"""

MIN_PRICE_INR = 500.0
MAX_PRICE_RATIO = 0.90          # Never charge more than 90% of dedicated cost
DEFAULT_FUEL_COST_PER_KM = 35.0 # ₹/km average diesel cost in India
DEFAULT_PLATFORM_FEE = 100.0    # ₹ flat platform fee


def calculate_price(
    base_trip_cost: float,
    truck_capacity_kg: float,
    shipment_weight_kg: float,
    detour_distance_km: float,
    fuel_cost_per_km: float = DEFAULT_FUEL_COST_PER_KM,
    platform_fee: float = DEFAULT_PLATFORM_FEE,
) -> dict:
    """
    Calculate the shared-capacity price for a shipment.

    Returns a dict with all breakdown components and the final price.
    """
    # Guard against zero-division
    if truck_capacity_kg <= 0:
        truck_capacity_kg = 10000.0

    # --- Volume Share ---
    # Proportional to how much of the truck the shipment uses
    weight_ratio = min(1.0, shipment_weight_kg / truck_capacity_kg)
    volume_share = base_trip_cost * weight_ratio

    # --- Detour Cost ---
    # Compensate the trucker for extra distance driven
    detour_cost = detour_distance_km * fuel_cost_per_km

    # --- Raw Final Price ---
    raw_price = volume_share + detour_cost + platform_fee

    # --- Safeguards ---
    final_price = max(MIN_PRICE_INR, raw_price)
    max_allowed = base_trip_cost * MAX_PRICE_RATIO
    final_price = min(final_price, max_allowed)

    # --- Dedicated truck cost (what shipper would pay alone) ---
    dedicated_truck_cost = base_trip_cost

    # --- Savings ---
    estimated_savings = dedicated_truck_cost - final_price
    savings_percent = (estimated_savings / dedicated_truck_cost * 100) if dedicated_truck_cost > 0 else 0

    # --- Explanation ---
    weight_pct = round(weight_ratio * 100, 1)
    explanation = (
        f"Your cargo occupies {weight_pct}% of the truck's capacity. "
        f"The pricing reuses the truck's existing route, so you only pay a "
        f"proportional share (₹{volume_share:,.0f}) plus ₹{detour_cost:,.0f} for "
        f"the {detour_distance_km:.1f} km additional detour and a ₹{platform_fee:,.0f} platform fee. "
        f"This is far cheaper than hiring a dedicated truck for ₹{dedicated_truck_cost:,.0f}."
    )

    return {
        "volumeShare": round(volume_share, 2),
        "detourCost": round(detour_cost, 2),
        "platformFee": round(platform_fee, 2),
        "finalPrice": round(final_price, 2),
        "dedicatedTruckCost": round(dedicated_truck_cost, 2),
        "estimatedSavings": round(estimated_savings, 2),
        "savingsPercent": round(savings_percent, 1),
        "weightRatio": round(weight_ratio, 4),
        "explanation": explanation,
    }
